pkg/gadgets/traceloop/tracer/bpf/traceloop.bpf.c

```diff
@@ -0,0 +1,454 @@
+	__uint(key_size, sizeof(u64));
+	__uint(value_size, sizeof(struct remembered_args));
+	__uint(max_entries, 1024);
+} probe_at_sys_exit SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(key_size, sizeof(u64));
+	__uint(value_size, sizeof(struct pt_regs));
+	__uint(max_entries, 1024);
+} regs_map SEC(".maps");
+
+static inline int skip_exit_probe(int nr) {
+	return !!(nr == __NR_exit ||
+		nr == __NR_exit_group ||
+		nr == __NR_rt_sigreturn);
+}
+
+/*
+ * Highly inspired from ksnoop.bpf.c:
+ * https://github.com/iovisor/bcc/blob/f90126bb3770ea1bdd915ff3b47e451c6dde5c40/libbpf-tools/ksnoop.bpf.c#L280
+ */
+static inline u64 get_arg(struct pt_regs *regs, int i)
+{
+	switch(i) {
+	case 1:
+		return PT_REGS_PARM1_CORE_SYSCALL(regs);
+	case 2:
+		return PT_REGS_PARM2_CORE_SYSCALL(regs);
+	case 3:
+		return PT_REGS_PARM3_CORE_SYSCALL(regs);
+	case 4:
+		return PT_REGS_PARM4_CORE_SYSCALL(regs);
+	case 5:
+		return PT_REGS_PARM5_CORE_SYSCALL(regs);
+	case 6:
+		return PT_REGS_PARM6_CORE_SYSCALL(regs);
+	default:
+		bpf_printk("There is no PT_REGS_PARM%d_SYSCALL macro, check the argument!\n", i);
+		return 0;
+	}
+}
+
+/*
+ * sys_enter is defined as:
+ * TP_PROTO(struct pt_regs *regs, long id)
+ * (https://elixir.bootlin.com/linux/v5.19/source/include/trace/events/syscalls.h#L20)
+ * So, ctx->args[0] contains a struct pt_regs and ctx->args[1] the syscall ID.
+ */
+SEC("raw_tracepoint/sys_enter")
+int ig_traceloop_e(struct bpf_raw_tracepoint_args *ctx)
+{
+	struct remembered_args remembered = {};
+	u64 pid = bpf_get_current_pid_tgid();
+	struct syscall_def_t *syscall_def;
+	/*
+	 * Initialize struct to empty to be sure all fields (even padding) are zeroed:
+	 * https://github.com/iovisor/bcc/issues/2623#issuecomment-560214481
+	 */
+	struct syscall_event_t sc = {};
+	u64 ts = bpf_ktime_get_ns();
+	struct task_struct *task;
+	u64 nr = ctx->args[1];
+	struct pt_regs *args;
+	void *perf_buffer;
+	u64 mntns_id;
+	int ret;
+	int i;
+
+	sc.timestamp = ts;
+	sc.cont_nr = 0;
+	sc.cpu = bpf_get_smp_processor_id();
+	sc.pid = pid >> 32;
+	sc.typ = SYSCALL_EVENT_TYPE_ENTER;
+	sc.id = nr;
+
+	remembered.timestamp = ts;
+	remembered.nr = nr;
+
+	syscall_def = bpf_map_lookup_elem(&syscalls, &nr);
+	/*
+	 * syscalls map contains definition for specific syscall like read or
+	 * write.
+	 * All others syscalls, like nanosleep, are not in this map because
+	 * their signature is not specific, in this case, we use the default
+	 * definition.
+	 */
+	if (syscall_def == NULL)
+		syscall_def = &default_definition;
+
+	task = (struct task_struct*)bpf_get_current_task();
+	mntns_id = (u64) BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
+
+	perf_buffer = bpf_map_lookup_elem(&map_of_perf_buffers, &mntns_id);
+	if (!perf_buffer)
+		return 0;
+
+	bpf_get_current_comm(sc.comm, sizeof(sc.comm));
+
+	ret = bpf_map_update_elem(&regs_map, &pid, &empty, BPF_NOEXIST);
+	if (ret) {
+		bpf_printk("enter: there should not be any pt_regs for key %lu: %d\n", pid, ret);
```

Comment:
I think we need to avoid bpf_printk() in production to avoid the big warning message in dmesg.

Ditto in other occurrences.